setup_cfg_fmt.py

```diff
@@ -316,6 +316,44 @@ def _is_ok_classifier(s: str) -> bool:
     return [s for s in classifiers if _is_ok_classifier(s)]
 
 
+def _imp_classifiers(setup_cfg: str) -> Optional[str]:
+    tox_config_to_names = {
+        'py': 'CPython',
+        'pypy': 'PyPy',
+        'jython': 'Jython',
+    }
+
+    cfg = configparser.ConfigParser()
+    cfg.read(setup_cfg)
+    classifiers = cfg.get('metadata', 'classifiers', fallback='')
+    tags = []
+
+    for classifier in classifiers.strip().splitlines():
+        if classifier.startswith(
+            'Programming Language :: Python :: Implementation :: ',
+        ):
+            tags.append(classifier.split(' :: ')[-1])
+
+    tox_ini = _adjacent_filename(setup_cfg, 'tox.ini')
+    if os.path.exists(tox_ini):
+        cfg = configparser.ConfigParser()
+        cfg.read(tox_ini)
+
+        envlist = cfg.get('tox', 'envlist', fallback='')
+        if envlist:
+            for env in envlist.split(','):
+                for key, tag in tox_config_to_names.items():
+                    if (key in env) and (tag not in tags):
+                        tags.append(tag)
```

Comment:
there's a bit of a problem here

`pypy3` contains `py` and so it's going to be marked as cpython even when it's not

I think the mapping should be done in priority order and we can do the following for factors:

- split them by `-`
- remove numbers
- check for an exact containment of one of the factors

for example py38-django2 => [py, django] => cpython (it contains py)

Reply:
does multi-factor env allow for something like `py38-pypy3`?

Reply:
I don't think so

Reply:
ok, so I should consider these general expressions:

- `pyNM-<name>PQ`
- `pypyN-<name>PQ`

(where N, M, P, Q are digits)
do you agree?

Reply:
yeah and the digits are optional as well